src/xenia/patcher/patcher.cc

```diff
@@ -0,0 +1,217 @@
+      case 1:
+        xe::store_and_swap(address, uint8_t(patch_data_entry.new_value_));
+        break;
+      case 2:
+        xe::store_and_swap(address, uint16_t(patch_data_entry.new_value_));
+        break;
+      case 4:
+        xe::store_and_swap(address, uint32_t(patch_data_entry.new_value_));
+        break;
+      case 8:
+        xe::store_and_swap(address, uint64_t(patch_data_entry.new_value_));
+        break;
+      default:
+        XELOGE("Unsupported patch allocation size");
+        break;
+    }
+    // Restore previous protection
+    heap->Protect(patch_data_entry.memory_address_,
+                  patch_data_entry.alloc_size_, old_address_protect);
+
+    if (!is_any_patch_applied_) {
+      is_any_patch_applied_ = true;
+    }
+  }
+}
+
+PatchFileEntry PatchingSystem::ReadPatchFile(
+    const std::filesystem::path& file_path) {
+  PatchFileEntry patchFile;
+  std::shared_ptr<cpptoml::table> patch_toml_fields;
+
+  try {
+    patch_toml_fields = cpptoml::parse_file(file_path.u8string());
+  } catch (...) {
+    XELOGE("Cannot load patch file: {}", file_path.u8string());
+    patchFile.title_id = -1;
+    return patchFile;
+  };
+
+  auto title_name = patch_toml_fields->get_as<std::string>("title_name");
+  auto title_id = patch_toml_fields->get_as<std::string>("title_id");
+  auto title_hash = patch_toml_fields->get_as<std::string>("hash");
+
+  patchFile.title_id = strtoul((*title_id).c_str(), NULL, 16);
+  patchFile.hash = strtoull((*title_hash).c_str(), NULL, 16);
+  patchFile.title_name = *title_name;
+
+  auto tarr = patch_toml_fields->get_table_array("patch");
+
+  for (auto table : *tarr) {
+    PatchInfoEntry patch = PatchInfoEntry();
+    auto patch_name = *table->get_as<std::string>("name");
+    auto patch_desc = *table->get_as<std::string>("desc");
+    auto patch_author = *table->get_as<std::string>("author");
+    auto is_enabled = *table->get_as<bool>("is_enabled");
+
+    patch.id = 0;  // ToDo: Implement id for future GUI stuff
+    patch.patch_name = patch_name;
+    patch.patch_desc = patch_desc;
+    patch.patch_author = patch_author;
+    patch.is_enabled = is_enabled;
+
+    const std::string data_types[] = {"be64", "be32", "be16", "be8"};
+
+    // Iterate through all available data sizes
+    for (const std::string& type : data_types) {
+      auto entries = ReadPatchData(type, table);
+
+      for (const PatchDataEntry& entry : *entries) {
+        patch.patch_data.push_back(entry);
+      }
+    }
+    patchFile.patch_info.push_back(patch);
+  }
+  return patchFile;
+}
+
+std::vector<PatchDataEntry>* PatchingSystem::ReadPatchData(
+    const std::string size_type,
+    const std::shared_ptr<cpptoml::table>& patch_table) {
+  std::vector<PatchDataEntry>* patch_data = new std::vector<PatchDataEntry>();
+  auto patch_data_tarr = patch_table->get_table_array(size_type);
+
+  if (!patch_data_tarr) {
+    return patch_data;
+  }
+
+  for (const auto& patch_data_table : *patch_data_tarr) {
```

Comment:
It'd probably be better if this loop were in the caller of this function and this function explicitly only cared about parsing individual patch entries.

It'd probably be OK if there was another function that did the handling of iterating the `data_types` array, I just think the iteration should be the responsibility of the outer scope.